File: Lab 6 (Sorting in Python)/insertionSort.py
import time
import random

def insertionSort(A, n):
    for i in range(1, n):
        temp = A[i]
        j = i - 1
        while j >= 0 and A[j] > temp:
            A[j + 1] = A[j]
            j = j - 1
        A[j + 1] = temp

def display(A, n):
    for i in range(n):
        print(A[i] , end=" ")
    print()  # to print new line after endign of loop

A = []
n = int(input("Enter n: "))
for i in range(n):
    A.append(random.randint(0, 100000))

start = time.time()
display(A, n)
insertionSort(A, n)
end = time.time()
display(A, n)
print(f"The time taken is {end - start:f} seconds.")

File: Lab 6 (Sorting in C)/insertionSort.c
// O(n^2)

#include<stdio.h>
#include<stdlib.h>
#include<time.h>

void insertionSort(int A[], int n) {
    int i, j, temp;
    for (i = 1; i < n; i++) {
        temp = A[i];
        j = i - 1;
        while (j >= 0 && A[j] > temp) {
            A[j + 1] = A[j];
            j = j - 1;
        }
        A[j + 1] = temp;
    }
}

void display(int A[], int n) {
    int i;
    for (i = 0; i < n; i++) {
        printf("%d ", A[i]);
    }
    printf("\n");
}

int main(){
    int A[100000], n, i;
    srand(time(NULL));
    clock_t start, end, diff;
    printf("Enter n: ");
    scanf("%d", &n);
    for (i = 0; i < n; i++)
        A[i] = rand();
    // display(A,n);
    start = clock();
    insertionSort(A,n);
    end = clock();
    // display(A,n);
    diff = end-start;
    printf("The time taken is %f seconds.\n", (double)diff/CLOCKS_PER_SEC);
    return 0;
}

//time taken for sorting 10 elements: 0.000010 s
//time taken for sorting 100 elements: 0.000016 s
//time taken for sorting 1000 elements: 0.002096 s
//time taken for sorting 10000 elements: 0.059925 s
//time taken for sorting 100000 elements: 3.184326 s